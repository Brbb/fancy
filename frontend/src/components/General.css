.langDropdown {
  font-size: 12px;
  height: 35px;
  min-width: 250px;
  border-radius: 10px;
  margin: 10px;
  box-shadow: 1px 1px 0px 2px rgba (0, 0, 0, 0.3);
  background: white;
  border: cornflowerblue solid 1px;
  color: cornflowerblue;
  cursor: pointer;
}

.generalSubmitBtn{
    margin-bottom: 20px
}

.radioGroup {
  width: 250px;
  margin-top: 5px;
  text-align: left;
}

.radioButton{
    font-size: 12px;
    margin-right: 10px
}
