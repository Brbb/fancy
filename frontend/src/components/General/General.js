import React, { Component } from "react";
import {
  Dropdown,
  RadioButtonGroup,
  Button,
  Message
} from "../Elements/Elements";
import { withRouter } from "react-router-dom";
import userApi from "../../services/users/api";

class General extends Component {
  constructor(props) {
    super(props);
    this.state = {
      outcome: { message: "", success: true },
      privacyOptions: [
        { text: "Private", value: "private" },
        { text: "Public", value: "public" }
      ]
    };
  }

  save = async () => {
    let result = await userApi.update(this.props.user);
    if (!result.err) {
      this.setState({ outcome: { message: "Settings saved!", success: true } });
    } else
      this.setState({
        outcome: {
          message: "Error while saving your settings.",
          success: false
        }
      });
  };

  handleLanguageSettingChange = event => {
    var editableUser = this.props.user;
    editableUser.settings.language = event.target.value;
    this.setState({ user: editableUser });
  };

  handlePrivacySettingChange = event => {
    var editableUser = this.props.user;
    editableUser.settings.isPrivate = event.target.value === "private";
    this.setState({ user: editableUser });
  };

  render() {
    return (
      <div className="f-section">
      <label className="f-section-title-no-margin">Preferences</label>
          <hr className="f-hr-section" />
        <Dropdown
          options={this.props.languages}
          text="Language"
          onChange={this.handleLanguageSettingChange}
          selected={this.props.user.settings.language}
        />

        <RadioButtonGroup
          options={this.state.privacyOptions}
          label="Privacy"
          checkedValue={
            this.props.user.settings.isPrivate ? "private" : "public"
          }
          onChange={this.handlePrivacySettingChange}
        />
        <Button
          className="f-submit-button"
          text="Save changes"
          onClick={this.save}
        />
        <Message
          text={this.state.outcome.message}
          success={this.state.outcome.success}
        />
      </div>
    );
  }
}

export default withRouter(General);
