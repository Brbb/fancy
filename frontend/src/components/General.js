import React, { Component } from "react";
import "./General.css";

const languages = [
  "English",
  "French",
  "German",
  "Japanese",
  "Chinese",
  "Korean",
  "Thai"
];

class Security extends Component {
  render() {
    return (
      <div>
        <form className="form" onSubmit={this.handleSubmit}>
          <label className="textFieldLabel">Language</label>
          <select className="langDropdown">
            {languages.map(language => {
              return (
                <option key={language} value={language}>
                  {language}
                </option>
              );
            })}
          </select>
          <label className="textFieldLabel">Privacy</label>
          <div className="radioGroup">
            <label className="radioButton">
              <input type="radio" name="gender" value="public" />
              Public
            </label>
            <label className="radioButton">
              <input type="radio" name="gender" value="private" />
              Private
            </label>
          </div>

          <input
            className="submitBtn generalSubmitBtn"
            type="button"
            value="Save changes"
            onClick={this.login}
          />
        </form>
      </div>
    );
  }
}

export default Security;
