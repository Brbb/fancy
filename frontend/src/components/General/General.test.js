import General from "./General";
import userApi from "../../services/users/api";

describe("General component", () => {
  let mockUser = {
    settings: {
      language: "Chinese",
      isPrivate: false
    },
    email: "asd",
    _id: "obj1234"
  };

  describe("Rendering tests", () => {
    test("General renders and assigns user settings correctly", () => {
      let wrapper = shallow(<General.WrappedComponent user={mockUser} />);
      expect(wrapper).toMatchSnapshot();

      let dropdown = wrapper.find("Dropdown");
      expect(dropdown.prop("selected")).toEqual(mockUser.settings.language);
      let radioButtonGroup = wrapper.find("RadioButtonGroup");
      expect(radioButtonGroup.prop("checkedValue")).toEqual(
        mockUser.settings.isPrivate ? "private" : "public"
      );
    });

    test("Select language updates settings correctly", () => {
      let wrapper = shallow(<General.WrappedComponent user={mockUser} />);
      expect(wrapper).toMatchSnapshot();

      jest.spyOn(userApi, "update").mockImplementation(() => {
        return Promise.resolve({});
      });

      let saveButton = wrapper.find("Button");
      let dropdown = wrapper.find("Dropdown");
      expect(dropdown.prop("selected")).toEqual(mockUser.settings.language);
      
      dropdown.simulate("change", { target: { value: "Thai" } }); // calling handleLanguageSettingChange
      
      expect(dropdown.prop("selected")).not.toEqual(mockUser.settings.language); // expecting 'Thai' for the user
      expect(userApi.update).toHaveBeenCalledTimes(0);

      return saveButton
        .prop("onClick")()
        .then(() => {
          expect(userApi.update).toHaveBeenCalledTimes(1);
          expect(wrapper.state().outcome.success).toBeTruthy();
          jest.clearAllMocks();
        });
    });

    test("Select privacy updates settings correctly", () => {
        let wrapper = shallow(<General.WrappedComponent user={mockUser} />);
        expect(wrapper).toMatchSnapshot();
  
        jest.spyOn(userApi, "update").mockImplementation(() => {
          return Promise.resolve({});
        });
  
        let saveButton = wrapper.find("Button");
        let radioButtonGroup = wrapper.find("RadioButtonGroup");
        expect(radioButtonGroup.prop("checkedValue")).toEqual('public');
        expect(mockUser.settings.isPrivate).toBeFalsy();
        
        radioButtonGroup.simulate("change", { target: { value: "private" } }); // calling handlePrivacySettingChange
        
        expect(mockUser.settings.isPrivate).toBeTruthy();
        expect(userApi.update).toHaveBeenCalledTimes(0);
  
        return saveButton
          .prop("onClick")()
          .then(() => {
            expect(userApi.update).toHaveBeenCalledTimes(1);
            expect(wrapper.state().outcome.success).toBeTruthy();
            jest.clearAllMocks();
          });
      });
  });
});
