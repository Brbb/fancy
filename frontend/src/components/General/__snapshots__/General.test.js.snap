// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`General component Rendering tests General renders and assigns user settings correctly 1`] = `
<div
  className="f-section"
>
  <label
    className="f-section-title-no-margin"
  >
    Preferences
  </label>
  <hr
    className="f-hr-section"
  />
  <Dropdown
    onChange={[Function]}
    selected="Chinese"
    text="Language"
  />
  <RadioButtonGroup
    checkedValue="public"
    label="Privacy"
    onChange={[Function]}
    options={
      Array [
        Object {
          "text": "Private",
          "value": "private",
        },
        Object {
          "text": "Public",
          "value": "public",
        },
      ]
    }
  />
  <Button
    className="f-submit-button"
    onClick={[Function]}
    text="Save changes"
  />
  <Message
    success={true}
    text=""
  />
</div>
`;

exports[`General component Rendering tests Select language updates settings correctly 1`] = `
<div
  className="f-section"
>
  <label
    className="f-section-title-no-margin"
  >
    Preferences
  </label>
  <hr
    className="f-hr-section"
  />
  <Dropdown
    onChange={[Function]}
    selected="Chinese"
    text="Language"
  />
  <RadioButtonGroup
    checkedValue="public"
    label="Privacy"
    onChange={[Function]}
    options={
      Array [
        Object {
          "text": "Private",
          "value": "private",
        },
        Object {
          "text": "Public",
          "value": "public",
        },
      ]
    }
  />
  <Button
    className="f-submit-button"
    onClick={[Function]}
    text="Save changes"
  />
  <Message
    success={true}
    text=""
  />
</div>
`;

exports[`General component Rendering tests Select privacy updates settings correctly 1`] = `
<div
  className="f-section"
>
  <label
    className="f-section-title-no-margin"
  >
    Preferences
  </label>
  <hr
    className="f-hr-section"
  />
  <Dropdown
    onChange={[Function]}
    selected="Thai"
    text="Language"
  />
  <RadioButtonGroup
    checkedValue="public"
    label="Privacy"
    onChange={[Function]}
    options={
      Array [
        Object {
          "text": "Private",
          "value": "private",
        },
        Object {
          "text": "Public",
          "value": "public",
        },
      ]
    }
  />
  <Button
    className="f-submit-button"
    onClick={[Function]}
    text="Save changes"
  />
  <Message
    success={true}
    text=""
  />
</div>
`;
